gapis/api/vulkan/framegraph.go

```diff
@@ -16,14 +16,153 @@ package vulkan
 
 import (
 	"context"
+	"fmt"
 
-	"github.com/google/gapid/core/log"
 	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/api/sync"
+	"github.com/google/gapid/gapis/capture"
+	"github.com/google/gapid/gapis/resolve/dependencygraph2"
 	"github.com/google/gapid/gapis/service/path"
 )
 
+// renderpassInfo stores a renderpass' info relevant for the framegraph.
+type renderpassInfo struct {
+	id       uint64
+	beginIdx api.SubCmdIdx
+	endIdx   api.SubCmdIdx
+	nodes    []dependencygraph2.NodeID
+	deps     map[uint64]struct{} // set of renderpasses this renderpass depends on
+}
+
 // GetFramegraph creates the framegraph of the given capture.
 func (API) GetFramegraph(ctx context.Context, p *path.Capture) (*api.Framegraph, error) {
-	log.W(ctx, "TODO: implement framegraph creation")
-	return nil, nil
+	config := dependencygraph2.DependencyGraphConfig{
+		SaveNodeAccesses:    true,
+		ReverseDependencies: true,
+	}
+	dependencyGraph, err := dependencygraph2.GetDependencyGraph(ctx, p, config)
+	if err != nil {
+		return nil, err
+	}
+
+	rpInfos := []*renderpassInfo{}
+	var rpInfo *renderpassInfo
+	currRpId := uint64(0)
+
+	// postSubCmdCb effectively processes each subcommand to extract renderpass info.
+	postSubCmdCb := func(state *api.GlobalState, subCmdIdx api.SubCmdIdx, cmd api.Cmd, i interface{}) {
```

Comment:
I can see that this is a form of closure that modifies the function fields several times in a loop in `MutateWithSubcommands` but this is quite subtle and can be misleading. 

This might be quite common in functional paradigm which I am not familiar with but as the rest of the code is quite non-functional(no puns intended) I think having a small helper struct with one function that actually does the operation and passing a lambda that calls the particular function in the object would be much cleaner.

Reply:
I think such closures are fairly common, but I agree with your comment nonetheless. I've extracted this lambda logic into a new `processSubCommand()` function.